### ISSUE TITLE
connection timeout removes all nodes from cluster

### ISSUE THREAD

**Issue by stephanos** [peillis/memcachir#10]
We noticed some odd behaviour in production: under certain circumstances, memcachir lost all nodes in our app. But they were running just fine.

Turned out, that if there is a _connection timeout_, memcachir thinks all nodes are gone:
https://github.com/peillis/memcachir/blob/562c8a69301f6f5b8288ed9524f0055bbde2d797/lib/memcachir/discovery/elasticache.ex#L16

cc @michaeljguarino

**Comment by stephanos** [comment]
@michaeljguarino I guess that never happened to you?

**Comment by michaeljguarino** [comment]
Oh I've never seen that.  We probably need to retry on that error.

**Comment by michaeljguarino** [comment]
What's the exact error tuple you're seeing?

**Comment by stephanos** [comment]
@michaeljguarino Yep, retrying would be good. And/or ignoring timeouts maybe, just in case the retries all fail? I've not actually seen the error tuple since it's not logged anywhere. Our best guess is that it's due to a timeout.

**Comment by stephanos** [comment]
Oh, I just notice that it _is_ logged. let me search the logs.

**Comment by stephanos** [comment]
I see `unable to fetch ElastiCache servers: :timeout` in the logs.

We rolled back to 3.2.1 for our services because of this. Seemed to do the trick. I think 3.2.1 just restarts itself for every _change_: A timeout would yield `[]` and that would be different from the existing setup, a restart would happen, next time the fetch would work again and another restarts puts us back to where we were. Not ideal, but it seems to work just fine.

What we observed with 3.2.2 was that _once_ there was a timeout talking to ElastiCache, all subsequent GET calls would error with `unable to get: no_nodes`. It seemed to be _stuck_. My colleague said the only option was to manually restart the app. I just glanced at the `herd` code to see what it does once it received `[]` as the node list result, but I couldn't figure out what it does exactly yet.

**Comment by michaeljguarino** [comment]
It shouldn't kill anything I believe, it should just dynamically supervise nothing then poll again and add them when something is discovered.

I can add a retry, and add a test that takes everything out then tries to add things back in to verify it works as it should.

**Comment by stephanos** [comment]
That would be awesome, thank you.

**Comment by stephanos** [comment]
I just talked to my colleague again and I think I know what's going on now. When we upgraded to 3.2.2 the `elasticachex` version we used was still 1.1.1 (missing the memory leak fix). Then, since we have lots of apps connect to that ElastiCache host, memcached must have run out of connections and so our apps were seeing timeouts. That's what triggered the problem. And since we never closed any connections, it looked like it was the client library's fault. 3.2.1 would have restarted in that case and closed all the connections.

Anyway, one way or another, I think a few retries might be good because it would make it more robust.

**Comment by michaeljguarino** [comment]
Ahh, yeah that makes sense.  I'll try to add some timeouts asap though, and dig into the failure to reregister everything.

**Comment by stephanos** [comment]
@michaeljguarino did you make any progress on the retries? last night our app seemed to have issues talking to ElasticCache (I see 36 `unable to fetch ElastiCache servers: :timeout`) and nodes were removed and added again for a while. It'd be much appreciated to make this more robust.

**Comment by stephanos** [comment]
my coworkers point out that it'd be even better if an error like `timeout` (even after retries), should not lead to the whole cluster being removed. it should just continue with what it has. I think that's a very fair point.

**Comment by stephanos** [comment]
please let me know whether you intend to fix this (I believe this requires some changes to `herd`). We'd prefer to move forward but might have to go back to 3.2.1 (which doesn't exhibit those kinds of issues).

**Comment by michaeljguarino** [comment]
Sorry, I got sidetracked.  I think I almost had it all fixed actually. Let me see if I can open a pr.

I'm not sure about implementing ignoring timeouts.  It can theoretically be done of course.

**Comment by michaeljguarino** [comment]
Here's the pr: https://github.com/peillis/memcachir/pull/12

**Comment by michaeljguarino** [comment]
On second thought, I'm actually wondering if the herd behavior should just ignore cluster's being emptied entirely, just to err on the side of availability.  I'll work on implementing that too after the pr is merged.

**Comment by stephanos** [comment]
> I'm actually wondering if the herd behavior should just ignore cluster's being emptied entirely, just to err on the side of availability

I think that'd be a good idea.
